packages/runtime/container-runtime/src/opCompressor.ts

```diff
@@ -0,0 +1,57 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { IsoBuffer } from "@fluidframework/common-utils";
+import { IBatchMessage } from "@fluidframework/container-definitions";
+import { compress } from "lz4js";
+import { BatchMessage } from "./batchManager";
+import { CompressionAlgorithms, ContainerRuntimeMessage } from "./containerRuntime";
+
+/**
+ * Compresses batches of ops. It generates a single compressed op that contains
+ * the contents of each op in the batch. It then submits empty ops for each original
+ * op to reserve sequence numbers.
+ */
+export class OpCompressor {
+    private compressedBatchCount = 0;
+
+    constructor(private readonly logger: ITelemetryLogger) {}
+
+    public compressBatch(batch: BatchMessage[], originalLength: number): IBatchMessage[] {
+        const batchToSend: IBatchMessage[] = [];
+        this.compressedBatchCount++;
+        const batchedContents: ContainerRuntimeMessage[] = [];
+
+        for (const message of batch) {
+            batchedContents.push(message.deserializedContent);
+        }
+
+        const compressionStart = Date.now();
+        const contentsAsBuffer = new TextEncoder().encode(JSON.stringify(batchedContents));
```

Comment:
It seems a bit expensive to start over with deserialized content and strigify it again.
Can we work with serialized content only somehow?
While it's not ideal (IMHO), I had somewhat similar problem [here](https://github.com/microsoft/FluidFramework/pull/12058/files#diff-4955d937755632e07ceb6cff6c215904ae4e7cc8ceacdbf801868e3188760594R507), where I concatenated strings into a single string, using \0 as a separator. I had one benefit you do not (I think) - I could build slower recovery code (though I know see that I got it wrong :) - will fix it).

Not sure what is the right solution here, but, if possible, we should avoid this expensive operation.

Reply:
I really wanted to avoid this stringification as well. Maybe I can build the JSON array manually to avoid doing this? Let me experiment with it a bit locally.

Reply:
JS environment sucks in this regard. It's possible that the best outcome here is
- convert each of the strings into array (using utf-8 encoding).
- Process each of them through compression algorithm. Ideally it supports streaming, i.e. you can feed data chunk by chunk to it (if not, you would need to copy all the data into bigger buffer).
   - One way or another, we need to somehow keep track of chunk lengths or add separators (like \0, similar in the example I mentioned below, which is not ideal).

It's probably Ok, but a bunch of code to write. I'd suggest measuring it. In example I mentioned above (with ODSP driver), we build a reproducible local test case and run through a bunch of experiments (only in Node though) to validate if any flow makes sense or not.

Feel free to leave it as is - I can try to fill in :)